overlord/devicestate/devicemgr.go

```diff
@@ -357,11 +424,106 @@ func getSerial(t *state.Task, privKey asserts.PrivateKey, device *auth.DeviceSta
 	return serial, nil
 }
 
+type serialRequestConfig struct {
+	requestIDURL     string
+	serialRequestURL string
+	headers          map[string]string
+	proposedSerial   string
+	body             []byte
+}
+
+func (cfg *serialRequestConfig) applyHeaders(req *http.Request) {
+	for k, v := range cfg.headers {
+		req.Header.Set(k, v)
+	}
+}
+
+func getFromTransaction(tr *configstate.Transaction, gadgetName, key string, v interface{}) error {
+	err := tr.Get(gadgetName, key, v)
+	if err != nil {
+		if _, ok := err.(*configstate.NoOptionError); !ok {
+			return err
+		}
+	}
+	return nil
+}
+
+func getSerialRequestConfig(t *state.Task) (*serialRequestConfig, error) {
+	gadgetInfo, err := snapstate.GadgetInfo(t.State())
+	if err != nil {
+		return nil, fmt.Errorf("cannot find gadget snap and its name: %v", err)
+	}
+	gadgetName := gadgetInfo.Name()
+
+	tr, _ := configstate.NewTransaction(t.State())
+	var svcURL string
+	err = getFromTransaction(tr, gadgetName, "registration-service-url", &svcURL)
```

Comment:
I suggest splitting these options in two namespaces, so that we can reuse for other needs.
Some suggestions, already accounting for the nesting we discussed:

 * device-service.url
 * registration.proposed-serial
 * registration.headers
 * registration.body